server/src/main/java/org/opensearch/tasks/TaskResourceTracker.java

```diff
@@ -0,0 +1,153 @@
+    private ThreadMXBean threadMXBean;
+    private long searchBytes;
+    private long responseOverheadBytes;
+
+    /**
+     * Private constructor to keep class Singleton
+     */
+    private TaskResourceTracker() {
+        taskMap = new ConcurrentHashMap<>();
+        threadMXBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+        searchBytes = 0;
+        responseOverheadBytes = 0;
+        overhead = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Method for returning static instance of class
+     *
+     * @return singleton instance for class
+     */
+    public static TaskResourceTracker getInstance() {
+        return instance;
+    }
+
+    public void registerTaskForTracking(long taskId, List<String> indices, ShardId shardId, String actionName) {
+        taskMap.put(new TaskInfoKey(taskId, indices, shardId, actionName), new ArrayList<>());
+    }
+
+    public void registerWorkerForTask(long taskId, long workerId, long cpuCurrent, long bytesCurrent, String threadpoolName) {
+        // TODO remove this after identifying cases where it can be true
+        if (taskMap.get(new TaskInfoKey(taskId)) == null) {
+            return;
+        }
+
+        TaskWorkerResourceUtilInfo taskWorkerResourceUtilInfo =
+            new TaskWorkerResourceUtilInfo(workerId, cpuCurrent, cpuCurrent, bytesCurrent, bytesCurrent,
+                true, threadpoolName);
+
+        taskMap.get(new TaskInfoKey(taskId)).add(taskWorkerResourceUtilInfo);
+    }
+
+    public void unregisterWorkerForTask(long taskId, long workerId, long cpuCurrent, long bytesCurrent) {
+        // This happens when task is unregistered first and then the runnable finishes
+        if (taskMap.get(new TaskInfoKey(taskId)) == null) {
+            return;
+        }
+        TaskWorkerResourceUtilInfo taskWorkerResourceUtilInfo = taskMap.get(new TaskInfoKey(taskId)).stream()
+            .filter(twrui -> twrui.getWorkerId() != null && twrui.getWorkerId() == workerId
+                && twrui.isActive()).findFirst().get();
+        taskWorkerResourceUtilInfo.setHeapNow(bytesCurrent);
+        taskWorkerResourceUtilInfo.setCpuNow(cpuCurrent);
+        taskWorkerResourceUtilInfo.setActive(false);
+    }
+
+    private void snapshot() {
+        final long[] search = {0};
+        final long[] response = {0};
+        taskMap.forEach((taskInfoKey, taskWorkerResourceUtilInfos) -> taskWorkerResourceUtilInfos.forEach(taskWorkerResourceUtilInfo -> {
+            if (taskWorkerResourceUtilInfo.isActive()) {
+                taskWorkerResourceUtilInfo.setHeapNow(threadMXBean.getThreadAllocatedBytes(taskWorkerResourceUtilInfo.getWorkerId()));
+            }
+            if ("tw".equals(taskWorkerResourceUtilInfo.getThreadPoolName())) {
+                response[0] += taskWorkerResourceUtilInfo.getOverheardBytes();
+            } else {
+                search[0] += taskWorkerResourceUtilInfo.getHeapNow() - taskWorkerResourceUtilInfo.getHeapStart();
```

Comment:
Use the optimized batch API instead
https://docs.oracle.com/javase/7/docs/jre/api/management/extension/com/sun/management/ThreadMXBean.html#getThreadAllocatedBytes(long[])